File: app/src/main/java/com/example/fooddeuk/activity/DetailRestaurantActivity.java
package com.example.fooddeuk.activity;

import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.View;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.TextView;

import com.example.fooddeuk.R;
import com.example.fooddeuk.adapter.ImageViewPager;
import com.example.fooddeuk.adapter.MenuListAdapter;
import com.example.fooddeuk.fragment.HomeFragment;
import com.example.fooddeuk.fragment.RestMenuCategoryFragment;
import com.example.fooddeuk.http.RestaurantService;
import com.example.fooddeuk.model.restaurant.RestaurantItem;
import com.orhanobut.logger.Logger;

import org.parceler.Parcels;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;
import me.relex.circleindicator.CircleIndicator;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class DetailRestaurantActivity extends AppCompatActivity implements MenuListAdapter.OnItemClickListener,MenuListAdapter.OnCartCountClickListener, View.OnClickListener {

    RestaurantItem.Restaurant restaurant;
    FragmentManager fragmentManager;
    FragmentTransaction fragmentTransaction;
    RestMenuCategoryFragment restMenuCategoryFragment;
    HomeFragment homeFragment;
    ImageViewPager imageViewPager;




//    @BindView(R.id.rest_detail_back_btn)
//    Button rest_detail_back_btn;


    @BindView(R.id.app_bar)
    AppBarLayout appBarLayout;
    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @BindView(R.id.rest_detail_back)
    ImageView rest_detail_back;
    @BindView(R.id.rest_detail_name)
    TextView rest_detai_name;
    @BindView(R.id.rest_detail_heart)
    ImageView rest_detail_heart;
    @BindView(R.id.rest_detail_cart)
    ImageView rest_detail_cart;

    @BindView(R.id.rest_detail_image_viewpager)
    ViewPager rest_detail_image_viewpager;

    @BindView(R.id.rest_detail_image_viewpager_indicator)
    CircleIndicator rest_detail_image_viewpager_indicator;

    @BindView(R.id.rest_main_tab)
    FrameLayout rest_main_tab;
//    @BindView(R.id.rest_detail_image)
//    ImageView rest_detail_image;
    //이름
//    @BindView(R.id.rest_detail_name)
//    TextView rest_detail_name;

    @BindView(R.id.toolbar_layout)
    CollapsingToolbarLayout collapsingToolbarLayout;


    //평점
    @BindView(R.id.rest_detail_rating)
    TextView rest_detail_rating;
    //리뷰숫자
    @BindView(R.id.rest_detail_review_count)
    TextView rest_detail_review_count;
    //단골숫자
    @BindView(R.id.rest_detail_dangol_count)
    TextView rest_detail_dangol_count;


    //별
//    @BindView(R.id.rest_detail_rating_star_layout)
//    LinearLayout rest_detail_rating_star_layout;

    @BindView(R.id.rest_detail_main_tab)
    TabLayout tabLayout;

    //TabLayout
//    @BindView(R.id.rest_detail_main_tab)
//    TabLayout rest_detail_main_tab;

    //Menu ViewPager
//    @BindView(R.id.rest_detail_main_tab_viewpager)
//    MainTabViewPager rest_detail_main_tab_viewpager;




    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_detail_restaurant);
        ButterKnife.bind(this);

        //Parcel로 선택된 식당 정보 가져옴
        /**from CurLocationRestaurantFragment**/
        /**from CurLocationRestaurantFragment**/
        restaurant = Parcels.unwrap(getIntent().getParcelableExtra("restaurant"));

        //메뉴관련 ViewPager 생성
        restMenuCategoryFragment=RestMenuCategoryFragment.newInstance(restaurant);

        //HomeFragment 생성
        homeFragment=new HomeFragment();
        //ImageViewPager 등록
        RestaurantService.getPicture(restaurant.rest_id).enqueue(new Callback<ArrayList<String>>() {
            @Override
            public void onResponse(Call<ArrayList<String>> call, Response<ArrayList<String>> response) {
                Logger.d(response.body());
                if(response.isSuccessful()){
                    imageViewPager=new ImageViewPager(DetailRestaurantActivity.this,response.body());
                    rest_detail_image_viewpager.setAdapter(imageViewPager);
                    rest_detail_image_viewpager_indicator.setViewPager(rest_detail_image_viewpager);
                }
            }

            @Override
            public void onFailure(Call<ArrayList<String>> call, Throwable t) {
                Logger.d("Ff");
            t.printStackTrace();
            }
        });





        viewSetting();

        fragmentManager = getSupportFragmentManager();
        fragmentTransaction = fragmentManager.beginTransaction();
        fragmentTransaction.add(R.id.rest_main_tab, restMenuCategoryFragment);
        fragmentTransaction.add(R.id.rest_main_tab, homeFragment);
        fragmentTransaction.hide(homeFragment);
        fragmentTransaction.commit();

        tabLayout.addOnTabSelectedListener(new TabLayout.OnTabSelectedListener() {
            @Override
            public void onTabSelected(TabLayout.Tab tab) {
// get the current selected tab's position and replace the fragment accordingly
                Fragment fragment = null;
                switch (tab.getPosition()) {
                    case 0:

                        fragmentManager = getSupportFragmentManager();
                        fragmentTransaction = fragmentManager.beginTransaction();
                        fragmentTransaction.show(restMenuCategoryFragment);
                        fragmentTransaction.hide(homeFragment);
//                        fragmentTransaction.replace(R.id.rest_main_tab, restMenuCategoryFragment);
                        fragmentTransaction.commit();
                        break;
                    case 1:
                        FragmentManager fm = getSupportFragmentManager();
                        FragmentTransaction ft = fm.beginTransaction();
//                        ft.replace(R.id.rest_main_tab, homeFragment);
                        ft.hide(restMenuCategoryFragment);
                        ft.show(homeFragment);
                        ft.commit();

                        break;
                    case 2:
                        fragmentManager = getSupportFragmentManager();
                        fragmentTransaction = fragmentManager.beginTransaction();
                        fragmentTransaction.replace(R.id.rest_main_tab, restMenuCategoryFragment);
                        fragmentTransaction.commit();
                        break;
                }

//                FragmentManager fm = getSupportFragmentManager();
//                FragmentTransaction ft = fm.beginTransaction();
//                ft.replace(R.id.rest_main_tab, fragment);
//                ft.commit();
            }

            @Override
            public void onTabUnselected(TabLayout.Tab tab) {
            }

            @Override
            public void onTabReselected(TabLayout.Tab tab) {
            }
        });
    }


    //Implement CartAddClick -> 메뉴 추가 한 뒤 장바구니 개수 새로고침
    @Override
    public void onCartCount(int cartSize) {
//        rest_detail_cart_count.setText(String.valueOf(cartSize));
    }

    //TODO 더 안정화
    //Implement ItemClick -> 해당 position의 상세 레이아웃 펼쳐짐
    @Override
    public void onItemClick(RecyclerView.ViewHolder view, float y,RecyclerView recyclerView) {
        Logger.d(view.itemView.getY());

    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
//            case R.id.rest_detail_back_btn:
//                finish();
//                break;

//

//            case R.id.detailTopCartButton:
//                IntentUtil.startActivity(this, CartActivity.class);
//                break;
//            case R.id.rest_detail_cart_btn:
//                IntentUtil.startActivity(this, CartActivity.class);
//                break;
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
//        rest_detail_cart_count.setText(String.valueOf(RealmUtil.getDataSize(CartItem.class)));

    }



    private void viewSetting() {
        //MenuList

        rest_detai_name.setText(restaurant.getName());
        rest_detai_name.setTextColor(Color.WHITE);
        setSupportActionBar(toolbar);
        final Drawable backArrow = getResources().getDrawable(R.drawable.ic_back_black);
        final Drawable heart = getResources().getDrawable(R.drawable.ic_heart);
        final Drawable cart=getResources().getDrawable(R.drawable.ic_cart);


        backArrow.setColorFilter(getResources().getColor(R.color.white), PorterDuff.Mode.SRC_ATOP);
        heart.setColorFilter(getResources().getColor(R.color.white), PorterDuff.Mode.SRC_ATOP);
        cart.setColorFilter(getResources().getColor(R.color.white), PorterDuff.Mode.SRC_ATOP);
        rest_detail_back.setImageDrawable(backArrow);
        rest_detail_back.setOnClickListener(view -> finish());

        rest_detail_heart.setImageDrawable(heart);
        rest_detail_cart.setImageDrawable(cart);
        appBarLayout.addOnOffsetChangedListener((appBarLayout, verticalOffset) -> {
            //collapse
            if (Math.abs(verticalOffset) == appBarLayout.getTotalScrollRange()) {
                backArrow.setColorFilter(getResources().getColor(R.color.charcoal_grey), PorterDuff.Mode.SRC_ATOP);
                heart.setColorFilter(getResources().getColor(R.color.charcoal_grey), PorterDuff.Mode.SRC_ATOP);
                cart.setColorFilter(getResources().getColor(R.color.charcoal_grey), PorterDuff.Mode.SRC_ATOP);
            } else if (verticalOffset == 0) {
                backArrow.setColorFilter(getResources().getColor(R.color.white), PorterDuff.Mode.SRC_ATOP);
                heart.setColorFilter(getResources().getColor(R.color.white), PorterDuff.Mode.SRC_ATOP);
                cart.setColorFilter(getResources().getColor(R.color.white), PorterDuff.Mode.SRC_ATOP);
            } else {
                // Somewhere in between
            }
//                int toolBarHeight = toolbar.getMeasuredHeight();
//                int appBarHeight = appBarLayout.getMeasuredHeight();
//                Float f = ((((float) appBarHeight - toolBarHeight) + verticalOffset) / ( (float) appBarHeight - toolBarHeight)) * 255;
////                toolbar.getBackground().setAlpha(255 - Math.round(f));
//                toolbar.setAlpha(255 - Math.round(f));
        });

//        Picasso.with(this).load(restaurant.rest_picture).fit().into(rest_detail_image);
        rest_detail_rating.setText(Double.toString(restaurant.getRating()));
        rest_detail_review_count.setText("평가("+restaurant.reviewCnt+")");
        rest_detail_dangol_count.setText("/  단골수 : "+restaurant.dangolCnt);
        collapsingToolbarLayout.setTitle(restaurant.name);
        collapsingToolbarLayout.setExpandedTitleColor(Color.WHITE);
        int bottomPx = 180; // margin in dips
        float d = getResources().getDisplayMetrics().density;
        int margin = (int)(bottomPx * d);
        collapsingToolbarLayout.setExpandedTitleMarginBottom(margin);
        collapsingToolbarLayout.setExpandedTitleTextAppearance(R.style.ExpandedAppBar);

//        rest_detail_back_btn.setOnClickListener(this);




    }


//    private void setStarView(float rating, LinearLayout rest_detail_rating_star_layout) {
//        for (int j = 0; j < 5; j++) {
//            ImageView star = new ImageView(getApplicationContext());
//            star.setLayoutParams(new ViewGroup.LayoutParams(starDpMap.get("width"), starDpMap.get("height")));
//            if (rating == 0) {
//                star.setImageResource(R.drawable.ic_star);
//                rest_detail_rating_star_layout.addView(star);
//
//            } else if (rating >= 1.0) {
//                rating -= 1.0;
//                star.setImageResource(R.drawable.ic_star_ranked);
//                rest_detail_rating_star_layout.addView(star);
//
//            } else if (rating == 0.5) {
//                rating -= 0.5;
//                star.setImageResource(R.drawable.ic_star_ranked_half);
//                rest_detail_rating_star_layout.addView(star);
//            }
//        }
//    }


//    public class fragmentPagerAdapter extends FragmentPagerAdapter {
//
//        public fragmentPagerAdapter(FragmentManager fm) {
//            super(fm);
//        }
//
////        @Override
////        public int getItemPosition(Object object) {
////            return POSITION_NONE;
////        }
//
//        @Override
//        public int getCount() {
//            return 3;
//        }
//
//
//        @Override
//        public Fragment getItem(int position) {
//            switch (position) {
//                case 0:
//                   return RestMenuCategoryFragment.newInstance(restaurant.rest_id);
//                case 1:
//                    return RestMenuCategoryFragment.newInstance(restaurant.rest_id);
//                case 2:
//                    return RestMenuCategoryFragment.newInstance(restaurant.rest_id);
//
//            }
//            return null;
//        }
//
//        @Override
//        public CharSequence getPageTitle(int position) {
//            switch (position) {
//                case 0:
//                    return "메뉴";
//                case 1:
//                    return "메뉴판";
//                case 2:
//                    return "정보";
//
//                default:return null;
//            }
//        }
//    }


}

File: app/src/main/java/com/example/fooddeuk/activity/DetailRestaurantActivity.kt
package com.example.fooddeuk.activity

import android.graphics.PorterDuff
import android.os.Bundle
import android.support.design.widget.TabLayout
import android.support.v4.content.ContextCompat
import android.support.v4.widget.NestedScrollView
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewTreeObserver
import com.example.fooddeuk.GlobalApplication
import com.example.fooddeuk.R
import com.example.fooddeuk.adapter.restaurantImageVPAdapter
import com.example.fooddeuk.fragment.RestMenuFragment
import com.example.fooddeuk.home.HomeFragment
import com.example.fooddeuk.listview.menu.MenuListAdapter
import com.example.fooddeuk.model.menu.MenuCategory
import com.example.fooddeuk.model.restaurant.Restaurant
import com.example.fooddeuk.network.HTTP.Single
import com.example.fooddeuk.rx.RxBus
import com.example.fooddeuk.util.StartActivity
import com.example.fooddeuk.util.addFragmentToActivity
import com.example.fooddeuk.util.replaceFragmentToActivity
import com.example.fooddeuk.util.toPx
import com.ogaclejapan.smarttablayout.SmartTabLayout
import com.orhanobut.logger.Logger
import com.trello.rxlifecycle2.android.lifecycle.kotlin.bindToLifecycle
import io.reactivex.functions.Consumer
import kotlinx.android.synthetic.main.activity_detail_restaurant.*


/**from NearRestaurantFragment */

class DetailRestaurantActivity : AppCompatActivity(), MenuListAdapter.OnItemClickListener, MenuListAdapter.OnCartCountClickListener, View.OnClickListener {
    var scrollFirstToolbar = 0
    var scrollSecondToolbar = 0 //->menu scroll base
    var menuItemHeight = 0
    var nameHeight = 0
    lateinit var restaurant: Restaurant
    lateinit var restMenuFragment: RestMenuFragment
    lateinit var homeFragment: HomeFragment
    lateinit var restaurantImageVPAdapter: restaurantImageVPAdapter
    lateinit var menuCategoryList: ArrayList<MenuCategory>
    lateinit var fakeTab : SmartTabLayout

    private val cart by lazy {
        ContextCompat.getDrawable(this, R.drawable.ic_cart)?.apply { setColorFilter(resources.getColor(R.color.white), PorterDuff.Mode.SRC_ATOP) }
    }
    private val heart by lazy {
        ContextCompat.getDrawable(this, R.drawable.ic_heart)?.apply { setColorFilter(resources.getColor(R.color.white), PorterDuff.Mode.SRC_ATOP) }
    }
    private val backArrow by lazy {
        ContextCompat.getDrawable(this, R.drawable.ic_back_black)?.apply { setColorFilter(resources.getColor(R.color.white), PorterDuff.Mode.SRC_ATOP) }
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_detail_restaurant)
        fakeTab=findViewById(R.id.tab_rest_menu_fake)

        val mGlobalLayoutListener = object : ViewTreeObserver.OnGlobalLayoutListener {
            override fun onGlobalLayout() {
                //TODO View가 다 그려진 뒤에 값 구하기
                //리스너삭제
                scroll_rest_detail.viewTreeObserver.removeOnGlobalLayoutListener(this)
                scrollFirstToolbar = vp_rest_detail_image.height - toolbar.height
                nameHeight=rest_detail_name2.measuredHeight + 24.toPx +scrollFirstToolbar
                scrollSecondToolbar = (frame_rest_main.y).toInt() - toolbar.height


            }
        }

        scroll_rest_detail.viewTreeObserver.addOnGlobalLayoutListener(mGlobalLayoutListener)

        menuItemHeight = resources.getDimensionPixelOffset(R.dimen.menu_item)
        RxBus.intentSubscribe(RxBus.DetailRestaurantActivityData, this.javaClass, Consumer { it ->
            if (it is Restaurant) {
                restaurant = it
                menuCategoryList = restaurant.menuCategory
                setFragment()
                viewSetting()
            }
        })


        with(tab_rest_main) {
            tabMode = TabLayout.MODE_FIXED
            tabGravity = TabLayout.GRAVITY_FILL
            addTab(this.newTab().setText("메뉴"), true)
            addTab(this.newTab().setText("스토리"))
            addTab(this.newTab().setText("후기"))
            setSelectedTabIndicatorHeight(0)
        }




//
//        tab_rest_main.addOnTabSelectedListener(object : TabLayout.OnTabSelectedListener {
//            override fun onTabSelected(tab: TabLayout.Tab) {
//                when (tab.position) {
//                    0 -> {
//                        showFragmentToActivity(restMenuFragment)
////                        hideFragmentToActivity(testFragment)
//                    }
//                    1 -> {
//                        hideFragmentToActivity(restMenuFragment)
////                        showFragmentToActivity(testFragment)
//                    }
//                    2 -> {
//                        hideFragmentToActivity(restMenuFragment)
////                        showFragmentToActivity(testFragment)
//                    }
//                }
//
//            }
//
//            override fun onTabUnselected(tab: TabLayout.Tab) {}
//
//            override fun onTabReselected(tab: TabLayout.Tab) {}
//        })
    }

    private fun setFragment() {
        restMenuFragment = RestMenuFragment.newInstance(restaurant)
        homeFragment = HomeFragment()
        addFragmentToActivity(R.id.frame_rest_main, restMenuFragment)
        addFragmentToActivity(R.id.frame_rest_main, homeFragment)
        replaceFragmentToActivity(R.id.frame_rest_main, restMenuFragment)

    }


    //Implement CartAddClick -> 메뉴 추가 한 뒤 장바구니 개수 새로고침
    override fun onCartCount(cartSize: Int) {
        //        rest_detail_cart_count.setText(String.valueOf(cartSize));
    }

    //TODO 더 안정화
    //Implement ItemClick -> 해당 position의 상세 레이아웃 펼쳐짐
    override fun onItemClick(view: RecyclerView.ViewHolder, y: Float, menuItemHeight: Int) {
        Logger.d(view.itemView.y)

    }

    override fun onClick(v: View) {
        when (v.id) {

        }//            case R.id.rest_detail_back_btn:
        //                finish();
        //                break;
        //
        //            case R.id.detailTopCartButton:
        //                IntentUtil.startActivity(this, CartActivity.class);
        //                break;
        //            case R.id.rest_detail_cart_btn:
        //                IntentUtil.startActivity(this, CartActivity.class);
        //                break;
    }

    override fun onResume() {
        super.onResume()
        //        rest_detail_cart_count.setText(String.valueOf(RealmUtil.getDataSize(CartItem.class)));

    }


    private fun viewSetting() {
        setSupportActionBar(toolbar)
        //식당 대표 사진
        Single(GlobalApplication.httpService.getPicture(restaurant._id)).bindToLifecycle(this).subscribe({
            restaurantImageVPAdapter = restaurantImageVPAdapter(this@DetailRestaurantActivity, it)
            vp_rest_detail_image.adapter = restaurantImageVPAdapter
            rest_detail_image_viewpager_indicator.setViewPager(vp_rest_detail_image)
        }, { it.printStackTrace() })
        rest_detail_name.text=restaurant.name
        rest_detail_name2.text = restaurant.name
        rest_detail_rating!!.text = java.lang.Double.toString(restaurant.getRating().toDouble()) + "(${restaurant.reviewCnt})"
        rest_detail_name.setTextColor(ContextCompat.getColor(this, R.color.white))
        rest_detail_back.setImageDrawable(backArrow)
        rest_detail_back.setOnClickListener { finish() }
        rest_detail_heart.setImageDrawable(heart)
        rest_detail_cart.setImageDrawable(cart)
        rest_detail_cart.setOnClickListener {
            RxBus.intentPublish(3, restaurant)
            StartActivity(SerachActivity::class.java)
        }

        scroll_rest_detail.setOnScrollChangeListener { v: NestedScrollView, scrollX: Int, scrollY: Int, oldScrollX: Int, oldScrollY: Int ->
//            logger("Y", scrollY)

            val toolbarIconAlpha = ((Math.min(1f, scrollY.toFloat() / scrollFirstToolbar)) * 255).toInt()
            val nameAlpha = (Math.min(1f, scrollY.toFloat() / nameHeight)).toInt()
            val tabAlpha = (Math.min(1f, scrollY.toFloat() / scrollSecondToolbar)).toInt()


            //매장이름
            if(nameAlpha==1){
                rest_detail_name.visibility=View.VISIBLE
            }else{
                rest_detail_name.visibility=View.GONE
            }

            //페이크 탭 레이아웃
            if (tabAlpha==1) {
                tab_rest_menu_fake.visibility = View.VISIBLE
            } else {
                tab_rest_menu_fake.visibility = View.INVISIBLE
            }
            //툴바 아이콘 칼러
            when (toolbarIconAlpha) {
                in 0..130 -> {
                    setIconChangeColor(R.color.white)
                }
                in 130..160 -> {
                    rest_detail_name.setTextColor(ContextCompat.getColor(applicationContext, R.color.white_1))
                    setIconChangeColor(R.color.white_1)
                }
                in 161..190 -> {
                    setIconChangeColor(R.color.white_2)
                }
                in 190..255 -> {
                    setIconChangeColor(R.color.black)
                }
            }

            toolbar.background.alpha = toolbarIconAlpha
        }


    }

    private fun setIconChangeColor(colorResource: Int) {
        backArrow?.setColorFilter(ContextCompat.getColor(this, colorResource), PorterDuff.Mode.SRC_ATOP)
        cart?.setColorFilter(ContextCompat.getColor(this, colorResource), PorterDuff.Mode.SRC_ATOP)
        heart?.setColorFilter(ContextCompat.getColor(this, colorResource), PorterDuff.Mode.SRC_ATOP)
    }

    override fun onDestroy() {
        RxBus.intentUnregister(this.javaClass)
        super.onDestroy()
    }


}
